### ISSUE TITLE
DataLoader `shuffle=False` ignored by DeepSpeedStrategy

### ISSUE THREAD

**Issue by FarzanT** [Lightning-AI/lightning#16772]
### Bug description

When a data loader with `shuffle=False` is given to a trainer that uses the DeepSpeedStrategy, the indices are all shuffled. A simple `print` statement in the `__get_item__` method of the DataSet clearly shows this. You must run the code below in a GPU enabled, non-interactive environment. (Google Colab is interactive)

### How to reproduce the bug

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.strategies import DeepSpeedStrategy


class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)

    def __getitem__(self, index):
        print("Given index is:", index)
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2, shuffle=False)

    model = BoringModel()
    trainer = Trainer(
        logger=False,
        default_root_dir=os.getcwd(),
        strategy=DeepSpeedStrategy(
        stage=1,
        pin_memory=True,
        ),
        accelerator="gpu",
        limit_train_batches=10,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data)


run()
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
* CUDA:
        - GPU:
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
                - Tesla V100-SXM2-32GB
        - available:         True
        - version:           11.4
* Lightning:
        - lightning-cloud:   0.5.26
        - lightning-utilities: 0.6.0.post0
        - pytorch-lightning: 1.9.1
        - reformer-pytorch:  1.4.4
        - torch:             1.13.0
        - torchmetrics:      0.11.0
        - torchtext:         0.14.0a0+f8d65dd
        - torchvision:       0.14.0
* Packages:
        - aiohttp:           3.8.3
        - aiosignal:         1.3.1
        - anndata:           0.8.0
        - anyio:             3.5.0
        - argon2-cffi:       21.3.0
        - argon2-cffi-bindings: 21.2.0
        - arrow:             1.2.3
        - asttokens:         2.0.5
        - async-timeout:     4.0.2
        - attrs:             22.1.0
        - av:                10.0.0
        - axial-positional-embedding: 0.2.1
        - backcall:          0.2.0
        - beautifulsoup4:    4.11.1
        - bleach:            4.1.0
        - blessed:           1.20.0
        - bottleneck:        1.3.5
        - brotlipy:          0.7.0
        - cached-property:   1.5.2
        - captum:            0.5.0
        - orjson:            3.8.6
        - packaging:         21.3
        - pandas:            1.5.2
        - pandocfilters:     1.5.0
        - parso:             0.8.3
        - pathtools:         0.1.2
        - patsy:             0.5.3
        - pexpect:           4.8.0
        - pickleshare:       0.7.5
        - pillow:            9.2.0
        - pip:               23.0
        - pkgutil-resolve-name: 1.3.10
        - product-key-memory: 0.1.10
        - prometheus-client: 0.14.1
        - promise:           2.3
        - prompt-toolkit:    3.0.20
        - protobuf:          3.19.1
        - psutil:            5.9.4
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - py-cpuinfo:        9.0.0
        - pycparser:         2.21
        - pydantic:          1.10.2
        - pygments:          2.14.0
        - pyjwt:             2.6.0
        - pynndescent:       0.5.8
        - pyopenssl:         22.0.0
        - pyparsing:         3.0.9
        - pyre-extensions:   0.0.23
        - pyrsistent:        0.18.0
        - pysocks:           1.7.1
        - python-dateutil:   2.8.2
        - python-dotenv:     0.21.1
        - python-editor:     1.0.4
        - python-multipart:  0.0.5
        - pytorch-lightning: 1.9.1
        - pytz:              2022.1
        - pyyaml:            6.0
        - pyzmq:             23.2.0
        - readchar:          4.0.3
        - reformer-pytorch:  1.4.4
        - requests:          2.28.1
        - rfc3986:           1.5.0
        - rich:              13.3.1
        - routing-transformer: 1.6.1
        - scanpy:            1.9.1
        - scikit-learn:      1.1.2
        - scipy:             1.8.1
        - seaborn:           0.12.1
        - send2trash:        1.8.0
        - sentencepiece:     0.1.96
        - sentry-sdk:        1.10.1
        - session-info:      1.0.0
        - setproctitle:      1.3.2
        - setuptools:        65.5.0
        - shortuuid:         1.0.10
        - six:               1.16.0
        - smmap:             5.0.0
        - sniffio:           1.2.0
        - soupsieve:         2.3.2.post1
        - stack-data:        0.2.0
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - statsmodels:       0.11.1
        - stdlib-list:       0.8.0
        - tabulate:          0.8.10
        - tensorboardx:      2.5.1
        - termcolor:         2.1.1
        - terminado:         0.13.1
        - threadpoolctl:     3.1.0
        - tinycss2:          1.2.1
        - torch:             1.13.0
        - torchmetrics:      0.11.0
        - torchtext:         0.14.0a0+f8d65dd
        - torchvision:       0.14.0
        - tornado:           6.1
        - tqdm:              4.64.1
        - traitlets:         5.9.0
        - typing-extensions: 4.3.0
        - typing-inspect:    0.8.0
        - ujson:             5.7.0
        - umap-learn:        0.5.2
        - urllib3:           1.26.12
        - uvicorn:           0.20.0
        - uvloop:            0.17.0
        - wandb:             0.13.5
        - watchfiles:        0.18.1
        - wcwidth:           0.2.5
        - webencodings:      0.5.1
        - websocket-client:  0.58.0
        - websockets:        10.4
        - wheel:             0.37.1
        - x-transformers:    1.5.1
        - xformers:          0.0.13
        - yarl:              1.8.2
        - zipp:              3.10.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         ppc64le
        - python:            3.8.13
        - version:           #1 SMP Thu Nov 17 08:57:02 EST 2022
```

</details>

### More info

Follows a discussion in Issue #15930. Pull Request #15931 was intended to fix this in general for DDP, but the changes seem to not apply to DeepSpeedStrategy. @awaelchli

**Comment by FarzanT** [comment]
Setting `replace_sampler_ddp=False` in `Trainer` resolves this issue.
